File: src/game/Maze.ts
import {
  PAC_CROSS_PAD,
  PAC_RADIUS,
  PAC_SCREEN_ANGLE,
  PRIZE_POINTS,
  angleDelta,
  gapHalfAngle,
  normAngle,
} from './constants'
import { fruitKindForLevel } from './leaderboard'
import type { Gap, Pellet, Prize, RadialWall } from './types'

function gapAt(mid: number, half: number): Gap {
  return { start: normAngle(mid - half), end: normAngle(mid + half) }
}

/**
 * Concentric circular maze.
 * Rings are Pac-Man lanes (0 = innermost playable, outward).
 * Radial spokes block travel along a ring so pellets sit in cul-de-sacs.
 */
export class Maze {
  readonly ringRadii: number[]
  readonly wallRadii: number[]
  readonly outerRadius: number
  readonly houseRadius: number
  readonly gaps: Gap[][]
  /** Spokes that block angular travel on a given ring. */
  spokes: RadialWall[] = []

  pellets: Pellet[] = []
  pelletCount = 0
  /** Total pellets at level start (for half-dot prize trigger). */
  pelletTotal = 0
  prize: Prize | null = null
  rotation = 0

  constructor() {
    this.houseRadius = 34
    this.wallRadii = [46, 78, 110, 142]
    this.outerRadius = 166
    this.ringRadii = [
      (this.wallRadii[0] + this.wallRadii[1]) / 2,
      (this.wallRadii[1] + this.wallRadii[2]) / 2,
      (this.wallRadii[2] + this.wallRadii[3]) / 2,
      (this.wallRadii[3] + this.outerRadius) / 2,
    ]

    const h0 = gapHalfAngle(this.wallRadii[0]) * 1.15
    const h1 = gapHalfAngle(this.wallRadii[1])
    const h2 = gapHalfAngle(this.wallRadii[2])
    const h3 = gapHalfAngle(this.wallRadii[3])

    this.gaps = [
      // Innermost wall / house gate — screen-fixed (does not rotate with the maze)
      [gapAt(-Math.PI / 2, h0)],
      [gapAt(1.1, h1), gapAt(1.1 + Math.PI, h1), gapAt(1.1 + Math.PI / 2, h1)],
      [
        gapAt(0.4, h2),
        gapAt(0.4 + (2 * Math.PI) / 3, h2),
        gapAt(0.4 + (4 * Math.PI) / 3, h2),
      ],
      [
        gapAt(1.85, h3),
        gapAt(1.85 + Math.PI, h3),
        gapAt(0.55, h3),
        gapAt(0.55 + Math.PI, h3),
      ],
    ]

    this.spokes = []
    this.rebuildSpokes(0)
    this.spawnPellets()
  }

  private spoke(ring: number, angle: number): RadialWall {
    const rInner = this.wallRadii[ring]
    const rOuter = ring === this.ringRadii.length - 1 ? this.outerRadius : this.wallRadii[ring + 1]
    return { ring, angle: normAngle(angle), rInner, rOuter }
  }

  /**
   * Place `count` radial walls (0–8). Angles are randomized per level but
   * always at least 1/8 turn (π/4) apart by picking distinct slots on an 8-point wheel.
   */
  rebuildSpokes(count: number): void {
    const n = Math.max(0, Math.min(8, Math.floor(count)))
    this.spokes = []
    if (n === 0) return

    const slotStep = (Math.PI * 2) / 8 // exactly 1/8 of the circle
    const base = Math.random() * Math.PI * 2
    const slots = [0, 1, 2, 3, 4, 5, 6, 7]
    // Fisher–Yates shuffle, take first n
    for (let i = slots.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1))
      ;[slots[i], slots[j]] = [slots[j], slots[i]]
    }
    for (let i = 0; i < n; i++) {
      const angle = base + slots[i] * slotStep
      const ring = Math.floor(Math.random() * this.ringRadii.length)
      this.spokes.push(this.spoke(ring, angle))
    }
  }

  /** Configure walls for a level and respawn pellets around them. */
  prepareLevel(wallCount: number): void {
    this.rebuildSpokes(wallCount)
    this.spawnPellets()
    this.prize = null
    this.rotation = 0
  }

  get ringCount(): number {
    return this.ringRadii.length
  }

  rotate(delta: number): void {
    this.rotation = normAngle(this.rotation + delta)
  }

  toLocal(screenAngle: number): number {
    return normAngle(screenAngle - this.rotation)
  }

  toScreen(localAngle: number): number {
    return normAngle(localAngle + this.rotation)
  }

  inGap(wallIndex: number, localAngle: number, pad = 0.012): boolean {
    // Wall 0 (house) is screen-fixed; convert local → screen before testing
    const a = normAngle(wallIndex === 0 ? this.toScreen(localAngle) : localAngle)
    for (const g of this.gaps[wallIndex]) {
      const start = normAngle(g.start - pad)
      const end = normAngle(g.end + pad)
      if (start <= end) {
        if (a >= start && a <= end) return true
      } else if (a >= start || a <= end) {
        return true
      }
    }
    return false
  }

  nearestGapMid(wallIndex: number, localAngle: number): number {
    // Compare in the gap's native space; always return a maze-local angle
    const query = wallIndex === 0 ? this.toScreen(localAngle) : normAngle(localAngle)
    let bestNative = query
    let bestDist = Infinity
    for (const g of this.gaps[wallIndex]) {
      const mid =
        g.start <= g.end
          ? normAngle((g.start + g.end) / 2)
          : normAngle(g.start + (Math.PI * 2 - g.start + g.end) / 2)
      let d = Math.abs(normAngle(mid) - normAngle(query))
      if (d > Math.PI) d = Math.PI * 2 - d
      if (d < bestDist) {
        bestDist = d
        bestNative = mid
      }
    }
    return wallIndex === 0 ? this.toLocal(bestNative) : bestNative
  }

  /** Screen-space midpoint of the house gate (always top of the dial). */
  houseGateScreenMid(): number {
    const g = this.gaps[0][0]
    if (g.start <= g.end) return normAngle((g.start + g.end) / 2)
    return normAngle(g.start + (Math.PI * 2 - g.start + g.end) / 2)
  }

  alignGapToPac(wallIndex: number, pacScreenAngle: number, localAngle: number): void {
    const mid = this.nearestGapMid(wallIndex, localAngle)
    this.rotation = normAngle(pacScreenAngle - mid)
  }

  canPacCross(fromRing: number, towardInner: boolean, localAngle: number): boolean {
    // Generous pad: accept up/down a little before Pac is dead-center in the hole
    if (towardInner) {
      if (fromRing <= 0) return false
      return this.inGap(fromRing, localAngle, PAC_CROSS_PAD)
    }
    if (fromRing >= this.ringCount - 1) return false
    return this.inGap(fromRing + 1, localAngle, PAC_CROSS_PAD)
  }

  wallIndexForPacCross(fromRing: number, towardInner: boolean): number {
    return towardInner ? fromRing : fromRing + 1
  }

  canGhostCross(fromRing: number, towardInner: boolean, localAngle: number): boolean {
    if (towardInner) {
      if (fromRing < 0) return false
      if (fromRing === 0) return this.inGap(0, localAngle)
      return this.inGap(fromRing, localAngle)
    }
    if (fromRing < 0) return this.inGap(0, localAngle)
    if (fromRing >= this.ringCount - 1) return false
    return this.inGap(fromRing + 1, localAngle)
  }

  ringRadius(ring: number): number {
    if (ring < 0) return this.houseRadius * 0.55
    return this.ringRadii[Math.max(0, Math.min(this.ringCount - 1, ring))]
  }

  wallIndexForCross(fromRing: number, towardInner: boolean): number {
    if (towardInner) {
      if (fromRing <= 0) return 0
      return fromRing
    }
    if (fromRing < 0) return 0
    return fromRing + 1
  }

  /**
   * Clamp maze rotation so Pac cannot slide through a radial spoke.
   * rotDelta > 0 → local angle decreases.
   */
  clampRotationForPac(ring: number, localAngle: number, rotDelta: number): number {
    return -this.clampAngularTravel(ring, localAngle, -rotDelta)
  }

  /** Clamp angular travel along a ring (positive = CCW / increasing angle). */
  clampAngularTravel(ring: number, localAngle: number, deltaLocal: number): number {
    if (ring < 0 || deltaLocal === 0) return deltaLocal
    const hitHalf = (PAC_RADIUS * 1.08) / this.ringRadius(ring)
    let allowed = deltaLocal

    for (const s of this.spokes) {
      if (s.ring !== ring) continue
      const toSpoke = angleDelta(localAngle, s.angle)
      if (deltaLocal > 0 && toSpoke > 0) {
        const max = Math.max(0, toSpoke - hitHalf)
        if (allowed > max) allowed = max
      } else if (deltaLocal < 0 && toSpoke < 0) {
        const max = Math.min(0, toSpoke + hitHalf)
        if (allowed < max) allowed = max
      }
    }
    return allowed
  }

  hitsSpoke(ring: number, localAngle: number, moveDir: 1 | -1): boolean {
    if (ring < 0) return false
    const hitHalf = (PAC_RADIUS * 1.08) / this.ringRadius(ring)
    for (const s of this.spokes) {
      if (s.ring !== ring) continue
      const toSpoke = angleDelta(localAngle, s.angle)
      if (Math.sign(toSpoke) === moveDir && Math.abs(toSpoke) <= hitHalf + 0.02) return true
    }
    return false
  }

  /** True if a spoke lies on the short arc from `from` toward `to` in direction `dir`. */
  spokeBlocksPath(ring: number, from: number, to: number, dir: 1 | -1): boolean {
    if (ring < 0) return false
    const span = Math.abs(angleDelta(from, to))
    for (const s of this.spokes) {
      if (s.ring !== ring) continue
      const d = angleDelta(from, s.angle)
      if (Math.sign(d) !== dir) continue
      if (Math.abs(d) > 0.001 && Math.abs(d) < span) return true
    }
    return false
  }

  spawnPellets(): void {
    this.pellets = []
    const outer = this.ringCount - 1
    const second = this.ringCount - 2
    const powerSpots: { ring: number; angle: number }[] = [
      { ring: outer, angle: 0 },
      { ring: outer, angle: Math.PI },
      { ring: second, angle: -Math.PI / 2 },
      { ring: second, angle: Math.PI / 2 },
    ]

    const counts = [16, 22, 28, 34]
    for (let ring = 0; ring < this.ringCount; ring++) {
      const n = counts[ring]
      for (let i = 0; i < n; i++) {
        const angle = (i / n) * Math.PI * 2
        const hitPower = powerSpots.some((p) => {
          if (p.ring !== ring) return false
          let d = Math.abs(normAngle(angle) - normAngle(p.angle))
          if (d > Math.PI) d = Math.PI * 2 - d
          return d < 0.12
        })
        if (hitPower) continue
        const onSpoke = this.spokes.some((s) => {
          if (s.ring !== ring) return false
          let d = Math.abs(normAngle(angle) - normAngle(s.angle))
          if (d > Math.PI) d = Math.PI * 2 - d
          return d < 0.1
        })
        if (onSpoke) continue
        this.pellets.push({ ring, angle, power: false, eaten: false })
      }
    }

    for (const p of powerSpots) {
      this.pellets.push({ ring: p.ring, angle: normAngle(p.angle), power: true, eaten: false })
    }

    this.pelletCount = this.pellets.length
    this.pelletTotal = this.pelletCount
    this.prize = null
  }

  resetPellets(): void {
    for (const p of this.pellets) p.eaten = false
    this.pelletCount = this.pellets.length
    this.pelletTotal = this.pelletCount
    this.prize = null
  }

  /** Pellets eaten so far this level. */
  get pelletsEaten(): number {
    return this.pelletTotal - this.pelletCount
  }

  /**
   * Spawn the level prize once on the bottom of the 3rd ring from outside
   * when at least half the dots are gone. Screen-fixed — does not rotate.
   */
  maybeSpawnPrize(level: number): void {
    if (this.prize) return
    if (this.pelletsEaten < this.pelletTotal * 0.5) return
    const ring = this.ringCount - 3 // 3rd from outside
    if (ring < 0) return
    const idx = Math.min(PRIZE_POINTS.length - 1, Math.max(0, level - 1))
    this.prize = {
      ring,
      angle: PAC_SCREEN_ANGLE, // screen-fixed at bottom of the dial
      points: PRIZE_POINTS[idx],
      kind: fruitKindForLevel(level),
      active: true,
    }
  }

  /** Fruit sits at screen-bottom; Pac eats it by being on that ring. */
  tryEatPrize(ring: number, _localAngle?: number): Prize | null {
    const p = this.prize
    if (!p || !p.active || p.ring !== ring) return null
    p.active = false
    this.prize = null
    return p
  }

  tryEat(ring: number, localAngle: number, radialSlop = 0.12): Pellet | null {
    for (const p of this.pellets) {
      if (p.eaten || p.ring !== ring) continue
      let d = Math.abs(normAngle(p.angle) - normAngle(localAngle))
      if (d > Math.PI) d = Math.PI * 2 - d
      if (d < radialSlop) {
        p.eaten = true
        this.pelletCount--
        return p
      }
    }
    return null
  }
}

File: native/OrbiPac/Game/Maze.swift
import Foundation
import CoreGraphics

final class Maze {
    let wallRadii: [CGFloat] = [46, 78, 110, 142]
    let outerRadius: CGFloat = 166, houseRadius: CGFloat = 34
    let ringRadii: [CGFloat]
    let gaps: [[Gap]]
    var spokes = [RadialWall](), pellets = [Pellet](), pelletCount = 0, pelletTotal = 0
    var prize: Prize?, rotation: CGFloat = 0

    init() {
        ringRadii = [(46+78)/2, (78+110)/2, (110+142)/2, (142+166)/2]
        func gap(_ mid: CGFloat, _ half: CGFloat) -> Gap { Gap(start: Constants.norm(mid-half), end: Constants.norm(mid+half)) }
        let h0 = Constants.gapHalfAngle(46) * 1.15, h1 = Constants.gapHalfAngle(78), h2 = Constants.gapHalfAngle(110), h3 = Constants.gapHalfAngle(142)
        // Three openings per playable wall, 120° apart. Adjacent walls stagger by 60°
        // so a cross never lands in the next hole without left/right travel.
        let base: CGFloat = 1.2, step = CGFloat.pi * 2 / 3, stagger = CGFloat.pi / 3
        gaps = [
            [gap(-.pi/2, h0)],
            [0,1,2].map { gap(base + CGFloat($0) * step, h1) },
            [0,1,2].map { gap(base + stagger + CGFloat($0) * step, h2) },
            [0,1,2].map { gap(base + CGFloat($0) * step, h3) },
        ]
        rebuildSpokes(0); spawnPellets()
    }
    var ringCount: Int { ringRadii.count }
    func rebuildSpokes(_ count: Int) {
        let n = max(0, min(8, count)); spokes = []; guard n > 0 else { return }
        let base = CGFloat.random(in: 0..<(.pi*2)); var slots = Array(0..<8); slots.shuffle()
        for slot in slots.prefix(n) {
            let ring = Int.random(in: 0..<ringCount), angle = Constants.norm(base + CGFloat(slot) * .pi*2/8)
            spokes.append(RadialWall(ring: ring, angle: angle, rInner: wallRadii[ring], rOuter: ring == ringCount-1 ? outerRadius : wallRadii[ring+1]))
        }
    }
    func prepareLevel(_ walls: Int) { rebuildSpokes(walls); spawnPellets(); prize=nil; rotation=0 }
    func rotate(_ delta: CGFloat) { rotation = Constants.norm(rotation + delta) }
    func toLocal(_ screen: CGFloat) -> CGFloat { Constants.norm(screen - rotation) }
    func toScreen(_ local: CGFloat) -> CGFloat { Constants.norm(local + rotation) }
    func inGap(_ wall: Int, _ local: CGFloat, _ pad: CGFloat = 0.012) -> Bool {
        let a = Constants.norm(wall == 0 ? toScreen(local) : local)
        return gaps[wall].contains { gap in
            let s=Constants.norm(gap.start-pad), e=Constants.norm(gap.end+pad)
            return s <= e ? (a >= s && a <= e) : (a >= s || a <= e)
        }
    }
    func nearestGapMid(_ wall: Int, _ local: CGFloat) -> CGFloat {
        let query = wall == 0 ? toScreen(local) : Constants.norm(local)
        let native = gaps[wall].min { Constants.angleDist(mid($0), query) < Constants.angleDist(mid($1), query) }!
        return wall == 0 ? toLocal(mid(native)) : mid(native)
    }
    private func mid(_ g: Gap) -> CGFloat { g.start <= g.end ? Constants.norm((g.start+g.end)/2) : Constants.norm(g.start + (.pi*2-g.start+g.end)/2) }
    func houseGateScreenMid() -> CGFloat { mid(gaps[0][0]) }
    func canPacCross(_ ring: Int, _ inner: Bool, _ a: CGFloat) -> Bool {
        inner ? (ring > 0 && inGap(ring,a,Constants.pacCrossPad)) : (ring < ringCount-1 && inGap(ring+1,a,Constants.pacCrossPad))
    }
    func wallIndexForPacCross(_ ring: Int, _ inner: Bool) -> Int { inner ? ring : ring+1 }
    func canGhostCross(_ ring: Int, _ inner: Bool, _ a: CGFloat) -> Bool {
        if inner { return ring >= 0 && inGap(ring == 0 ? 0 : ring,a) }
        return ring < 0 ? inGap(0,a) : (ring < ringCount-1 && inGap(ring+1,a))
    }
    func ringRadius(_ ring: Int) -> CGFloat { ring < 0 ? houseRadius * 0.55 : ringRadii[max(0,min(ringCount-1,ring))] }
    func wallIndexForCross(_ ring: Int, _ inner: Bool) -> Int { inner ? (ring <= 0 ? 0 : ring) : (ring < 0 ? 0 : ring+1) }
    func clampRotationForPac(_ ring:Int,_ a:CGFloat,_ d:CGFloat)->CGFloat { -clampAngularTravel(ring,a,-d) }
    func clampAngularTravel(_ ring: Int, _ a: CGFloat, _ delta: CGFloat) -> CGFloat {
        guard ring >= 0, delta != 0 else { return delta }; let half = Constants.pacRadius*1.08/ringRadius(ring); var allowed=delta
        for s in spokes where s.ring == ring { let d=Constants.delta(a,s.angle); if delta > 0 && d > 0 { allowed=min(allowed,max(0,d-half)) }; if delta < 0 && d < 0 { allowed=max(allowed,min(0,d+half)) } }
        return allowed
    }
    func spokeBlocksPath(_ ring:Int,_ from:CGFloat,_ to:CGFloat,_ dir:Int)->Bool {
        guard ring >= 0 else { return false }; let span=abs(Constants.delta(from,to))
        return spokes.contains { s in s.ring == ring && (Constants.delta(from,s.angle) > 0 ? 1 : -1) == dir && abs(Constants.delta(from,s.angle)) > 0.001 && abs(Constants.delta(from,s.angle)) < span }
    }
    func spawnPellets() {
        pellets=[]; let outer=ringCount-1, second=ringCount-2
        let powers=[(outer,CGFloat(0)),(outer,CGFloat.pi),(second,-CGFloat.pi/2),(second,CGFloat.pi/2)], counts=[16,22,28,34]
        for ring in 0..<ringCount { for i in 0..<counts[ring] { let a=CGFloat(i)/CGFloat(counts[ring]) * .pi*2
            if powers.contains(where: {$0.0==ring && Constants.angleDist(a,$0.1)<0.12}) || spokes.contains(where: {$0.ring==ring && Constants.angleDist(a,$0.angle)<0.1}) { continue }
            pellets.append(Pellet(ring:ring,angle:a,power:false,eaten:false))
        }}
        pellets += powers.map { Pellet(ring:$0.0,angle:Constants.norm($0.1),power:true,eaten:false,flip:false) }
        // One rare flip-dial pellet on an inner ring (mirrors controls + confuses ghosts).
        let flipRing = 1
        let flipAngle = Constants.norm(CGFloat.pi * 0.65)
        pellets.removeAll { !$0.power && $0.ring == flipRing && Constants.angleDist($0.angle, flipAngle) < 0.15 }
        pellets.append(Pellet(ring: flipRing, angle: flipAngle, power: false, eaten: false, flip: true))
        pelletCount=pellets.count; pelletTotal=pelletCount; prize=nil
    }
    func resetPellets() { for i in pellets.indices { pellets[i].eaten=false }; pelletCount=pellets.count; pelletTotal=pelletCount; prize=nil }
    var pelletsEaten: Int { pelletTotal-pelletCount }
    func maybeSpawnPrize(_ level:Int) {
        guard prize == nil, pelletsEaten >= pelletTotal/2 else { return }
        let ring = ringCount - 3
        guard ring >= 0 else { return }
        // 5% chance: Zookeeper's net instead of fruit.
        if CGFloat.random(in: 0..<1) < Constants.netSpawnChance {
            prize = Prize(ring: ring, angle: Constants.pacScreenAngle, points: 0, kind: -1, active: true, isNet: true)
            return
        }
        let idx = max(0, min(Constants.prizePoints.count - 1, level - 1))
        prize = Prize(
            ring: ring,
            angle: Constants.pacScreenAngle,
            points: Constants.prizePoints[idx],
            kind: fruitKindForLevel(level),
            active: true,
            isNet: false
        )
    }
    func tryEatPrize(_ ring:Int)->Prize? { guard let p=prize,p.active,p.ring==ring else{return nil}; prize=nil; return p }
    func tryEat(_ ring:Int,_ a:CGFloat,_ slop:CGFloat = 0.12)->Pellet? {
        tryEat(ring, a, slop: slop, regular: true, power: true, flip: true)
    }

    /// Eat matching pellet kinds. Mirror mode uses power/flip only while restoring regular dots.
    func tryEat(_ ring:Int,_ a:CGFloat, slop:CGFloat = 0.12, regular:Bool, power:Bool, flip:Bool)->Pellet? {
        for i in pellets.indices where !pellets[i].eaten && pellets[i].ring == ring && Constants.angleDist(pellets[i].angle,a) < slop {
            let p = pellets[i]
            if p.flip { guard flip else { continue } }
            else if p.power { guard power else { continue } }
            else { guard regular else { continue } }
            pellets[i].eaten=true; pelletCount -= 1; return pellets[i]
        }
        return nil
    }

    /// Mirror shadow: put a regular eaten dot back on the field.
    @discardableResult
    func tryRestoreDot(_ ring:Int,_ a:CGFloat,_ slop:CGFloat = 0.12)->Bool {
        for i in pellets.indices where pellets[i].eaten && !pellets[i].power && !pellets[i].flip
            && pellets[i].ring == ring && Constants.angleDist(pellets[i].angle,a) < slop {
            pellets[i].eaten = false
            pelletCount += 1
            return true
        }
        return false
    }
}
